src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php

```diff
@@ -186,4 +217,164 @@ private function assertProductCanBeUpdated(
+        );
+
+        // Better check with price tax excluded since it's the one saved in database, if the price matches
+        // the product's one no need for specific price
+        if ($priceTaxExcluded->equals(new Number((string) $initialProductPriceTaxExcl))) {
+            return;
+        }
+
+        $specificPrice = new SpecificPrice();
+        $specificPrice->id_shop = 0;
+        $specificPrice->id_cart = 0;
+        $specificPrice->id_shop_group = 0;
+        $specificPrice->id_currency = $order->id_currency;
+        $specificPrice->id_country = 0;
+        $specificPrice->id_group = 0;
+        $specificPrice->id_customer = $order->id_customer;
+        $specificPrice->id_product = $product->id;
+        $specificPrice->id_product_attribute = $combination ? $combination->id : 0;
+        $specificPrice->price = (float) (string) $priceTaxExcluded;
+        $specificPrice->from_quantity = 1;
+        $specificPrice->reduction = 0;
+        $specificPrice->reduction_type = 'amount';
+        $specificPrice->reduction_tax = !$priceTaxIncluded->equals($priceTaxExcluded);
+        $specificPrice->from = '0000-00-00 00:00:00';
+        $specificPrice->to = '0000-00-00 00:00:00';
+        $specificPrice->add();
+        $this->temporarySpecificPrices[] = $specificPrice;
+    }
+
+    /**
+     * @param Order $order
+     *
+     * @return Cart
+     */
+    private function createNewOrEditExistingCart(Order $order)
+    {
+        $cartId = Cart::getCartIdByOrderId($order->id);
+        if ($cartId) {
+            $cart = new Cart($cartId);
+        } else {
+            $cart = new Cart();
+            $cart->id_shop_group = $order->id_shop_group;
+            $cart->id_shop = $order->id_shop;
+            $cart->id_customer = $order->id_customer;
+            $cart->id_carrier = $order->id_carrier;
+            $cart->id_address_delivery = $order->id_address_delivery;
+            $cart->id_address_invoice = $order->id_address_invoice;
+            $cart->id_currency = $order->id_currency;
+            $cart->id_lang = $order->id_lang;
+            $cart->secure_key = $order->secure_key;
+
+            $cart->add();
+        }
+
+        Context::getContext()->cart = $cart;
+
+        return $cart;
+    }
+
+    /**
+     * @param int $combinationId
+     *
+     * @return Combination|null
+     */
+    private function getCombination($combinationId)
+    {
+        $combination = null;
+
+        if (0 !== $combinationId) {
+            $combination = new Combination($combinationId);
+
+            if (!Validate::isLoadedObject($combination)) {
+                throw new OrderException('Product combination not found.');
+            }
+        }
+
+        return $combination;
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param int $langId
+     *
+     * @return Product
+     */
+    private function getProductObject(ProductId $productId, $langId)
```

Comment:
There's no need to use a `ProductId` here, especially since it is created on the fly only for the purpose of calling this method (L115)